### PR TITLE
unrelated_type_equality_checks: ignore nullability

### LINKED ISSUE #2196: unrelated_type_equality_checks false positive in NNBD
In the following code `p` field is nullable but `p` param is non-nullable.

```dart
// @dart = 2.9
class A {
  int? p;
  bool m(num p) => p == this.p; // false positive here
}
```

If I make the `p` field non-nullable the lint disappear.

### PR DESCRIPTION
**Pull request by srawlins**
# Description

Nullability should be ignored when considering "relatedness". It is valid to compare `num` and `int?` objects.

Fixes #2196

### CODE REVIEW

File: lib/src/util/dart_type_utilities.dart [hunk 1]
```diff
@@ -396,40 +396,41 @@ class DartTypeUtilities {
...
-  /// * Two equal types are considered related, e.g. classes `int` and `int`,
-  ///   classes `List<String>` and `List<String>`,
-  ///   classes `List<T>` and `List<T>`, and type variables `A` and `A`.
-  /// * Two types such that one is a subtype of the other, such as classes
-  ///   `List<dynamic>` and `Iterable<dynamic>`, and type variables `A` and `B`
-  ///   where `A extends B`.
-  /// * Two types, each representing a class:
+  /// * Two types which are equal modulo nullability are considered related,
+  ///   e.g. `int` and `int`, `String` and `String?`, `List<String>` and
+  ///   `List<String>`, `List<T>` and `List<T>`, and type variables `A` and `A`.
```
Comment (bwilkerson) [inline lib/src/util/dart_type_utilities.dart:405]: Was one of these `List<String>`s suppose to have a question mark?

File: lib/src/util/dart_type_utilities.dart [hunk 2]
```diff
@@ -396,40 +396,41 @@ class DartTypeUtilities {
...
   ///   * are related if their supertypes are equal, e.g. `List<dynamic>` and
   ///     `Set<dynamic>`.
-  /// * Two types, each representing a type variable, are related if their
-  ///   bounds are related.
+  /// * Two type variables are related if their bounds are related.
   /// * Otherwise, the types are related.
   // TODO(srawlins): typedefs and functions in general.
   static bool unrelatedTypes(
-      TypeSystem typeSystem, DartType leftType, DartType rightType) {
+      TypeSystem typeSystem, DartType leftType_, DartType rightType_) {
```
Comment (bwilkerson) [inline lib/src/util/dart_type_utilities.dart:421]: Personally I'd prefer to not use underscore as a suffix, but to use names like `promotedLeftType` (or even just `left`) below, but that's clearly a matter of personal taste, so take it or leave it as you see fit.

**Review by bwilkerson — APPROVED** [review]

File: lib/src/util/dart_type_utilities.dart (hunk 1 above)
Comment (srawlins) [reply lib/src/util/dart_type_utilities.dart:405]: No. It didn't before.I just wanted to add one example where they are equal except for nullability.

File: lib/src/util/dart_type_utilities.dart (hunk 2 above)
Comment (srawlins) [reply lib/src/util/dart_type_utilities.dart:421]: I like it. Done.

**Pull request merged**